lib/content/write.js

```diff
@@ -44,101 +45,98 @@ function write (cache, data, opts) {
 
 module.exports.stream = writeStream
 
-function writeStream (cache, opts) {
-  opts = opts || {}
-  const inputStream = new PassThrough()
-  let inputErr = false
+// writes proxied to the 'inputStream' that is passed to the Promise
+// 'end' is deferred until content is handled.
+class CacacheWriteStream extends Flush {
+  constructor (cache, opts) {
+    super()
+    this.opts = opts
+    this.cache = cache
+    this.inputStream = new Minipass()
+    this.inputStream.on('error', er => this.emit('error', er))
+    this.handleContentP = null
+    // we don't actually get data, so don't buffer 'end' forever
+    // this.resume()
+  }
 
-  function errCheck () {
-    if (inputErr) {
-      throw inputErr
+  write (chunk, encoding, cb) {
+    if (!this.handleContentP) {
+      this.handleContentP = handleContent(
+        this.inputStream,
+        this.cache,
+        this.opts
+      )
     }
+    return this.inputStream.write(chunk, encoding, cb)
   }
 
-  let allDone
-  const ret = to(
-    (c, n, cb) => {
-      if (!allDone) {
-        allDone = handleContent(inputStream, cache, opts, errCheck)
+  flush (cb) {
+    this.inputStream.end(() => {
+      if (!this.handleContentP) {
+        const e = new Error('Cache input stream was empty')
+        e.code = 'ENODATA'
+        // empty streams are probably emitting end right away.
+        // defer this one tick by rejecting a promise on it.
+        return Promise.reject(e).catch(cb)
```

Comment:
🤔 I noticed there's a missing `stream.emit('error', e)` for this "Cache empty" error here, is that expected to be part of the public API? is it fine if it's just not there anymore? OR is the promise rejection going to trigger that somehow upstream?

Reply:
`minipass-flush` streams will emit an error if the flush cb is called with an error.  Since Minipass streams can emit `end` immediately when `end()` is called, and that would mean that the caller doesn't get a chance to catch the error, it's deferred by the `Promise.reject(e).catch(cb)` rather than doing `return cb(er)` synchronously.  I guess I could have also done `Promise.resolve(e).then(cb)`, but both are slightly faster than `setTimeout(() => cb(er))` or `setImmediate(() => cb(er))`.

Reply:
This line will cause hte `cb` to be called with an error as its first argument.  [minipass-flush](http://npm.im/minipass-flush) streams will emit an error if the `flush(cb)` callback is called with an error.

Reply:
I see! could be nice to have a note or example of that in `minipass-flush` README - btw I can't find it's source if you I can point out I can send a PR myself 😊